Catalog2 3

Mono1it opened the pull request:
Выбранная командой

архитектура: MVVM;
способ вёрстки: UIKit.
Ссылка на таск трекер: https://github.com/users/zanpakt0/projects/1/views/6

Сам файл .md с декомпозицией:
[Catalog_ShcherbakovIlya.md](https://github.com/user-attachments/files/23917810/Catalog_ShcherbakovIlya.md)

Первоначальная оценка времени выполнения задач и реальное время, потраченное на решении каждой задачи:
Модуль 2:
Верстка экрана коллекции NFT
- Сверстать изображение, заголовок и описании (est: 1.5 часа; fact: 1.5 часа).
- Сверстать коллекцию (est: 1 час; fact: 3 часа).
- Сверстать ячейку коллекции (est: 4 часа; fact: 4 часа).

Логика экрана каталога
- Настройка навигации между экранами (est: 1 час; fact: 1 час).

Логика экрана коллекции NFT
- Логика добавления в избранное (est: 2 часа; fact: 30 минут).
- Логика добавления/удаления из корзины (est: 2 часа; fact: 30 минут).
- Логика подстановки данных в изображение, заголовок и описание (est: 1 час; fact: 30 минут).
- Логика открытия экрана коллекции NFT (est: 1 час; fact: 20 минут).

Данные в ячейках коллекции и работа с сетевыми данными
- GET запрос картинки, имени и количества NFT для экрана каталога (est: 2 часа; fact: 2 часа).
- GET запрос картинки, заголовка и описания (est: 2 часа; fact: 1 час).
- МОК данные для ячеек коллекции (est: 1 час; fact: 30 минут).

Mukhammed003 committed:
develope(newFeature): Добавлен метод для отображения алерта с двумя кнопками в ErrorView

Выполнены следующие работы в UserCardViewController:
1) Обновлён Colors с добавлением forViewBackgound для установки заднего фона для WebViewVeiwController;
2) Обновлён Assets с корректировкой sortButton;
3) Добавлен метод showErrorAlertWithTwoButtons в ErrorView для показа алерта с двумя кнопками;
4) Обновлён WebViewViewController для правильного отображения в тёмной теме изменён:
1) background у view; 2) tintColor у navigationBar; 3) bottomAnchor у webView

Mukhammed003 committed:
Merge pull request #136 from zanpakt0/errorAlertWithTwoButtons

develope(newFeature): Добавлен метод для отображения алерта с двумя к…

Mono1it committed:
fix(sortCatalogButton): изменил название ассета

Mono1it committed:
Merge branch 'develop' of github.com:zanpakt0/iOS-FakeNFT-cohort-32-team-1 into catalog2-3

# Conflicts:
#	FakeNFT/Resources/Assets.xcassets/sortButton.imageset/Light 1.png
#	FakeNFT/Resources/Assets.xcassets/sortButton.imageset/Light.png

Mono1it committed:
fix(): Добавил потерявшиеся ассеты

Mono1it committed:
feat(NFTCollectionViewController): Сверстал изображение, заголовок и описание
feat(CatalogVC): Добавил модальный переход на экран nft колекции

Mono1it committed:
feat(NFTCollectionViewController): Сверстал коллекцию и добавил весь ui в scrollView
feat(NFTCell): Создал временную ячейку для настроки коллекции

Mono1it committed:
feat(NFTCollectionViewController): Добавил МОК данные, задал в коллекции размеры ячеек
feat(NFTCell): Сверстал ячейку для коллекции с двумя кнопками
feat(SimpleRatingView): Добавил хелпер для создания вью рейтинга
feat(CatalogStructures): Добал структуры данных для будующих запросов

Mono1it committed:
feat(CatalogProvider): реализовал логику GET запроса картинки, имени и количества NFT для экрана каталога

Mono1it committed:
feat(NFTCollectionViewMode): Реализовал логику ViewModel и сделал простой провайдер для коллекции nft
feat(NFTCollectionVC): Наладил навигацию между экранами NFTCollectionVC и NFTCollectionVM по архитектуре MVVM

Mono1it committed:
feat(NFTCollectionVC): Реализовал подстановку данных в в изображение, заголовок и описание

Mono1it committed:
Обновил файл с декомпозицией эпика "Каталог"

Mukhammed003 commented on FakeNFT/Scenes /Catalog/NFTCollection/NFTCell.swift:
```diff
@@ -0,0 +1,178 @@
+            // nftImageView
+            nftImageView.topAnchor.constraint(equalTo: contentView.topAnchor),
+            nftImageView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor),
+            nftImageView.heightAnchor.constraint(equalToConstant: 108),
```
Давай добавим все магические числа сделаем статичными для улучшения читаемости. 

К примеру: 
`enum NFTCellLayout {
    
    static let nftImageViewSizes: CGFloat = 108

}`

Mukhammed003 commented on FakeNFT/Scenes /Catalog/Catalog/CatalogViewController.swift:
```diff
@@ -3,13 +3,17 @@ import Combine
+        catalogProvider: CatalogProvider(
+            networkClient: DefaultNetworkClient(),
+            storage: CatalogStorageImpl())
+    )
```
Давай добавим инъекцию зависимостей чтобы улучшить гибкость и поддерживаемость кода.

К примеру через init: 

```
init(viewModel: CatalogViewModel) {
        self.viewModel = viewModel
        super.init(nibName: nil, bundle: nil)
    }
```

Mukhammed003 commented on FakeNFT/Scenes /Catalog/NFTCollection/NFTCell.swift:
```diff
@@ -0,0 +1,178 @@
+            nftImageView.topAnchor.constraint(equalTo: contentView.topAnchor),
+            nftImageView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor),
+            nftImageView.heightAnchor.constraint(equalToConstant: 108),
+            nftImageView.widthAnchor.constraint(equalToConstant: 108),
```
Магические числа в констрейнтах можно вынести в отдельный enum и сделать статичными. 

К примеру: 
```
enum NFTCellLayout {
    static let nftImageViewSizes: CGFloat = 108
}
```

Mukhammed003 commented on FakeNFT/Scenes /Catalog/NFTCollection/NFTCollectionViewController.swift:
```diff
@@ -0,0 +1,358 @@
+        let image = UIImageView()
+        image.translatesAutoresizingMaskIntoConstraints = false
+        image.image = UIImage(named: "Peach")
+        image.layer.cornerRadius = 12
```
Размеры для отдельных вью тоже можно вынести в отдельный enum как в примере с констрейнтами.

Mukhammed003 commented on FakeNFT/Scenes /Catalog/Catalog/CatalogViewController.swift:
```diff
@@ -132,10 +136,18 @@ extension CatalogViewController: UITableViewDataSource {
+        let NFTvc = NFTCollectionViewController(catalogItem: item)
+        NFTvc.modalTransitionStyle = .crossDissolve
+        NFTvc.modalPresentationStyle = .fullScreen
+        present(NFTvc, animated: true)
```
Можно открывать с помощью pushViewController чтобы не создавать кастомный вью для кнопки назад в NFTCollectionViewController: 
```
navigationController?.pushViewController(userCardViewController, animated: false)
```

Mukhammed003 commented on FakeNFT/Scenes /Catalog/NFTCollection/NFTCollectionViewModel.swift:
```diff
@@ -0,0 +1,33 @@
+        for id in nftIds {
+            provider.loadNft(id: id) { [weak self] result in
+                guard let self else { return }
+                self.isLoading = false
```
Мне кажется тут можно улучшить логику с помощью DispatchGroup(), чтобы isLoading был true пока не загрузятся все NFT. 

```
func loadData(nftIds: [String]) {
    guard !nftIds.isEmpty else { return }
    isLoading = true
    var loadedNFTs: [Nft] = []
    let group = DispatchGroup()

    for id in nftIds {
        group.enter()
        provider.loadNft(id: id) { result in
            defer { group.leave() }
            ...
            
    group.notify(queue: .main) { [weak self] in
        self?.nfts = loadedNFTs
        self?.isLoading = false
    }
}
```

Mukhammed003 commented on FakeNFT/Scenes /Catalog/NFTCollection/NFTCollectionViewController.swift:
```diff
@@ -0,0 +1,358 @@
+        return collection
+    }()
+    
+    private lazy var loadingIndicator: UIActivityIndicatorView = {
+        let indicator = UIActivityIndicatorView()
+        indicator.translatesAutoresizingMaskIntoConstraints = false
+        return indicator
+    }()
```
Для лоадера можно установить стиль .large: 
```
indicator.style = .large
```

Mono1it replied:
По макету размеры лоадера, вроде как, стандартные

Mukhammed003 commented on FakeNFT/Scenes /Catalog/NFTCollection/NFTCollectionViewController.swift:
```diff
@@ -0,0 +1,358 @@
+    private let catalogItem: Catalog
+    
+    private var favorites: [String] = []
+    private var itemsInCart: [String] = []
```
Давай выведем их в viewModel, так как по архитектуре MVVM всё что не касающееся UI отображения не должно храниться в viewController.

Mukhammed003 commented on FakeNFT/Scenes /Catalog/NFTCollection/NFTCollectionViewController.swift:
```diff
@@ -0,0 +1,358 @@
+            nftIds: catalogItem.nfts
+        )
+        super.init(nibName: nil, bundle: nil)
+    }
```
Лучше viewModel создавать снаружи чтобы не нарушать принцип инъекции зависимостей.

Mukhammed003 commented:
Дополнительно при иницализации catalogTabBarItem в TabBarController, я бы порекомендовал использовать системную иконку, так как для вкладки "Каталог" есть системная иконка: "square.stack.fill". 

Пример использования: 
```
UITabBarItem(
        title: NSLocalizedString("Tab.catalog", comment: ""),
        image: UIImage(systemName: "square.stack.fill"),
        tag: 0
    )
```

Mono1it committed:
refactor(Catalog): Исправил замечания от Мухаммеда(сокомандник)

Mono1it committed:
fix(CatalogVC): Создал viewModel снаружи, чтобы не нарушать принцип иньекции зависимостей

TimKosihinYT commented on FakeNFT/Scenes /Catalog/Catalog/CatalogViewModel.swift:
```diff
@@ -0,0 +1,42 @@
+    //MARK: - Init
+    init(catalogProvider: CatalogProvider) {
+        self.catalogProvider = catalogProvider
+        self.isLoading = true
```
Лучше сделать это в loadData

TimKosihinYT commented on FakeNFT/Scenes /Catalog/NFTCollection/NFTCell.swift:
```diff
@@ -0,0 +1,204 @@
+    
+    weak var delegate: NFTCellDelegate?
+    
+    private var inCart: Bool = false
```
Здесь и ниже можно опустить Bool и оставить просто false

TimKosihinYT commented on FakeNFT/Scenes /Catalog/NFTCollection/NFTCollectionViewController.swift:
```diff
@@ -0,0 +1,388 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = .background
+        updateCollectionViewHeight()
```
Можно тоже отправить в setupUI

TimKosihinYT reviewed (commented):
Оставил пару стилистических комментариев

Mono1it committed:
Исправил замечания ревьювера
stlyle(Catalog): Добавил поддержку черной темы
feat(NFTCollectionVC): Добавил UIRefreshControl для scrollView, который обновляет список коллекций
fix(NFTCollectionVM): Изменил метод load(), чтобы можно было его использовать для UIRefreshControl
